src/network/p2p/kad_rocksdb_store.rs

```diff
@@ -0,0 +1,238 @@
+fn ttl(expires: Instant) -> u32 {
+	(expires - Instant::now())
+		.max(Duration::from_secs(1))
+		.as_secs() as u32
+}
+
+impl From<kad::Record> for Entry {
+	fn from(record: kad::Record) -> Self {
+		Entry(
+			record.key.to_vec(),
+			KadRecord {
+				value: record.value,
+				publisher: record.publisher.map(PeerId::to_bytes).unwrap_or_default(),
+				ttl: record.expires.map(ttl).unwrap_or(0),
+			},
+		)
+	}
+}
+
+impl From<Entry> for kad::Record {
+	fn from(entry: Entry) -> Self {
+		let Entry(key, record) = entry;
+
+		kad::Record {
+			key: RecordKey::from(key),
+			value: record.value,
+			publisher: (!record.publisher.is_empty())
+				.then(|| PeerId::from_bytes(&record.publisher).expect("Invalid peer ID")),
+			expires: (record.ttl > 0)
+				.then(|| Instant::now() + Duration::from_secs(record.ttl.into())),
+		}
+	}
+}
+
+/// RocksDB implementation of a `RecordStore`.
+/// Providers are kept in memory.
+pub struct RocksDBStore {
+	/// The identity of the peer owning the store.
+	local_key: KBucketKey<PeerId>,
+	/// The configuration of the store.
+	config: RocksDBStoreConfig,
+	/// The stored (regular) records.
+	records: Arc<rocksdb::DB>,
+	/// The stored provider records.
+	providers: Providers,
+}
+
+/// Configuration for a `RocksDBStore`.
+/// There is no limit for maximum number of records like in MemoryStore.
+#[derive(Debug, Clone)]
+pub struct RocksDBStoreConfig {
+	/// The maximum size of record values, in bytes.
+	pub max_value_bytes: usize,
+	pub providers: ProvidersConfig,
+}
+
+impl Default for RocksDBStoreConfig {
+	// Default values kept in line with libp2p
+	fn default() -> Self {
+		Self {
+			max_value_bytes: 65 * 1024,
+			providers: Default::default(),
+		}
+	}
+}
+
+#[cfg(feature = "kademlia-rocksdb")]
+impl RocksDBStore {
+	/// Creates a new `RocksDBRecordStore` with the given configuration.
+	pub fn with_config(local_id: PeerId, config: RocksDBStoreConfig, db: Arc<rocksdb::DB>) -> Self {
+		RocksDBStore {
+			local_key: KBucketKey::from(local_id),
+			records: db,
+			providers: Providers::with_config(config.providers.clone()),
+			config,
+		}
+	}
+
+	/// Retains the records satisfying a predicate.
+	#[instrument(level = "trace", skip(self, f))]
+	pub fn retain<F>(&mut self, mut f: F)
+	where
+		F: FnMut(&RecordKey, &Record) -> bool,
+	{
+		let mut write_batch = WriteBatch::default();
+
+		self.records()
+			.filter(|record| !f(&record.key, record))
+			.for_each(|record| write_batch.delete(record.key.clone()));
+
+		let write_batch_len = write_batch.len();
+		match self.records.write(write_batch) {
+			Err(error) => error!("Failed to retain records that satisfies the predicate: {error}"),
+			Ok(_) => info!("Removed {write_batch_len} records from the RocksDB store"),
+		}
+	}
+
+	// Optimizations are not implemented currently
+	pub fn shrink_hashmap(&mut self) {}
+}
+
+impl RocksDBStore {
+	pub fn get_cf(&self) -> Option<Arc<BoundColumnFamily>> {
```

Comment:
We should prob instrument this function too, as it could introduce some performance penalties on frequent calls.